### PR TITLE
Add job controller test verifying if backoff is reseted on success

### PR DESCRIPTION
**Pull request by soltysh**
This adds a test for our job controller proving that the backoff limit is being reseted after a successful run. 

Fixes https://github.com/kubernetes/kubernetes/issues/54904

```release-note
NONE
```

### CODE REVIEW

**Comment by soltysh** [comment]
@kubernetes/sig-apps-pr-reviews

File: pkg/controller/job/job_controller_test.go [hunk 1]
```diff
@@ -1269,3 +1258,78 @@ func bumpResourceVersion(obj metav1.Object) {
...
+				{0, 0, 1, 0},
+			},
+		},
+		"parallelism=2 (just failure)": {
+			2, 2, 1,
+			[]pods{
+				{0, 2, 0, 1},
+				{0, 0, 1, 0},
+			},
+		},
```
Comment (soltysh) [inline pkg/controller/job/job_controller_test.go]: @foxish if I add the 3rd test here, with +1 failure and +1 success this is where it won't get reseted, and I'm not 100% convinced it should. /cc @clamoriniere1A

Comment (soltysh) [reply pkg/controller/job/job_controller_test.go]: My reasoning is that with small parallelism it should not be a problem, but what about big ones, esp. bigger than backoff limit? Or maybe we should ensure the backoff limit can't be lower than the parallelism, at least.

Comment (foxish) [reply pkg/controller/job/job_controller_test.go]: That makes sense, maybe we should set the `backofflimit` to `min(6, parallelism)`?

Comment (foxish) [reply pkg/controller/job/job_controller_test.go]: On second thought - maybe that's independent. With both failure and success at the same time, maybe we should leave the backoff time unchanged and retry?

Comment (soltysh) [reply pkg/controller/job/job_controller_test.go]: I've created https://github.com/kubernetes/kubernetes/issues/56853 to discuss this and have it solved for 1.10. Here I've added a separate commit to change the completions to 3 in that e2e to get rid of that flake.

**Comment by soltysh** [comment]
@foxish ptal

File: pkg/controller/job/job_controller_test.go (hunk 1 above)
Comment (soltysh) [reply pkg/controller/job/job_controller_test.go]: I want to leave that unit test for the backoff limit resetting, in place.

**Comment by foxish** [comment]
Thanks @soltysh. The test will be useful for checking the behavior in future.

**Comment by foxish** [comment]
Can you update the title of this PR @soltysh

**Comment by foxish** [comment]
/hold

@soltysh, these seem like real failures and not flakes.

**Comment by soltysh** [comment]
> @soltysh, these seem like real failures and not flakes.

Checking...

**Comment by soltysh** [comment]
OMG, I've dropped the completions on job definition but not on the expected line. And `framework.WaitForJobFinish` expects there a number explicitly. Updated should be good to go.

**Comment by soltysh** [comment]
The test are green I'm gonna manually re-apply lgtm label back so that when queue is open this gets in.

**Pull request merged**